src/query/api/v1/middleware/rewrite.go

```diff
@@ -0,0 +1,202 @@
+				base.ServeHTTP(w, r)
+				return
+			}
+
+			logger := opts.InstrumentOpts.Logger()
+			if err := rewriteRangeDuration(r, mwOpts, logger); err != nil {
+				logger.Warn("could not rewrite range. "+
+					"continuing with request", zap.Error(err))
+			}
+			base.ServeHTTP(w, r)
+		})
+	}
+}
+
+const (
+	queryParam = "query"
+	endParam   = "end"
+	startParam = "start"
+)
+
+func rewriteRangeDuration(
+	r *http.Request,
+	opts PrometheusRangeRewriteOptions,
+	logger *zap.Logger,
+) error {
+	// Extract relevant query params
+	query, start, end, err := extractParams(r)
+	if err != nil {
+		return err
+	}
+
+	// Query for namespace metadata of namespaces used to service the request
+	store := opts.Storage
+	ctx, fetchOpts, err := opts.FetchOptionsBuilder.NewFetchOptions(r.Context(), r)
+	if err != nil {
+		return err
+	}
+
+	attrs, err := store.QueryStorageMetadataAttributes(ctx, start, end, fetchOpts)
+	if err != nil {
+		return err
+	}
+
+	// Find the largest resolution
+	var res time.Duration
+	for _, attr := range attrs {
+		if attr.Resolution > res {
+			res = attr.Resolution
+		}
+	}
+
+	// Largest resolution is 0 which means we're routing to the unaggregated namespace.
+	// Unaggregated namespace can service all requests, so return.
+	if res == 0 {
+		return nil
+	}
+
+	// Rewrite ranges within the query, if necessary
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return err
+	}
+	updated := rewriteRangeInQuery(expr, res, opts.ResolutionMultiplier)
+	if !updated {
+		return nil
+	}
+
+	// Add updated query to the request where necessary
+	updatedQuery := expr.String()
+
+	// Update query param in URL, if present
+	urlQueryValues, err := url.ParseQuery(r.URL.RawQuery)
+	if err != nil {
+		return err
+	}
+	if urlQueryValues.Get(queryParam) != "" {
+		urlQueryValues.Set(queryParam, updatedQuery)
+	}
+	updatedURL, err := url.Parse(r.URL.String())
+	if err != nil {
+		return err
+	}
+	updatedURL.RawQuery = urlQueryValues.Encode()
+	r.URL = updatedURL
+
+	// Update query param in the request body, if present
+	if r.Form.Get(queryParam) != "" {
+		r.Form.Set(queryParam, updatedQuery)
+	}
+
+	// Reset the body on the request for any handlers that may want access to the raw body.
+	updatedBody := r.Form.Encode()
+	r.Body = ioutil.NopCloser(bytes.NewBufferString(updatedBody))
+
+	logger.Debug("rewrote range duration value within query",
+		zap.String("originalQuery", query),
+		zap.String("updatedQuery", updatedQuery))
+
+	return nil
+}
+
+func extractParams(r *http.Request) (string, time.Time, time.Time, error) {
+	if err := r.ParseForm(); err != nil {
+		return "", time.Time{}, time.Time{}, err
+	}
+
+	query := r.FormValue(queryParam)
+
+	now := time.Now()
+	start, err := prometheus.ParseTime(r, startParam, now)
+	if err != nil {
+		return "", time.Time{}, time.Time{}, err
+	}
+
+	end, err := prometheus.ParseTime(r, endParam, now)
+	if err != nil {
+		return "", time.Time{}, time.Time{}, err
+	}
+	if start.After(end) {
```

Comment:
is the idea that this would result in a validation error, so there's nothing to do?

Reply:
Ah no, this is a bug! Should be generating an error here. Good catch